core/sdk/ui: L1 canonical bridge option

kangarang opened the pull request:
this pr adds support for sending assets via L2-native L1 canonical bridges

![Screen Shot 2022-02-05 at 1 31 22 AM](https://user-images.githubusercontent.com/18040654/152631460-8901842b-9c6d-4789-875e-509519134db2.png)

kangarang committed:
sdk: add estimateGas and populateTx methods

- CanonicalBridge: `populateDepositTx`, `estimateDepositTx`

kangarang committed:
core/sdk/frontend: update l1CanonicalBridge addresses

kangarang committed:
frontend: add L1 canonical bridge option

increase refetchInterval to 10 seconds

kangarang committed:
core/sdk/frontend cleanup:

- update deps

- mv liquidity check to hook

- configure correct addresses and interfaces

- check connected network before checking allowance

- logic for setting l1 canonical bridge

- add l1 canonical bridge hook

- update mainnet.json

kangarang committed:
frontend: use radio buttons to select native bridge

group tooltip components together

shanefontaine reviewed (changes requested):
Nice! This is going to be a solid feature.

* It looks like the Arbitrum bridge that the tx is sent to is not the actual address that gets approved. See attached image. It looks like the approval address for sending DAI over Arbitrum is `0xd3b5b60020504bc3489d6949d545893982ba3011`, and I don't see that in the repo anywhere. I did not look at any other bridges, but it might be worth confirming that the approval addresses are accounted for if they differ from the actual bridge address that is interacted with

![Screen Shot 2022-02-05 at 14 42 27](https://user-images.githubusercontent.com/9441295/152661548-1419bfe7-1fa8-406e-9b3a-9bf3a34d87d7.png)

* I thought we landed on keeping the canonical bridge logic out of the SDK and keeping the SDK pure with Hop-only logic. I see that the code is pretty closely coupled, so I'd imagine that might be difficult. Do you remember where we ended up landing on that last time we discussed this?
  * Similar comment but WRT the canonical bridge addresses in `core/src/addresses/`

* I see a hardCoded `gasLimit` of `200e3` [here](https://github.com/hop-protocol/hop/pull/388/files#diff-8d5518358c199a7acaf429850259ed0b559b724895d70fca003708ab0479d3cdR101) (and maybe other places):
  * I wouldn't think we would ever need to hard code a `gasLimit`. If that is true, can it be removed?
  * Arbitrum has a different concept of gas than the other chains. Because of this their gasLimit values are much, much higher — higher than `200e3` for most or all txs. I’d imagine the `200e3` might not work for Arbiturm. See these equivalent txs gas costs on [Optimism](https://optimistic.etherscan.io/tx/0x95378c74e0af381364ec32508b7cfca8bbb5572dc8d4e2c4b5df3e7e42cd12d4 ) and [Arbitrum](https://arbiscan.io/tx/0x9d4a0885dd5d8c21909a934b3b2ea98949c577a42e24319c3b5fafbd455a8ac0).
  * I will caveat this comment with the knowledge that there are Alchemy bugs on both Mainnet and Polygon WRT gas limit stuff

kangarang commented on packages/frontend/src/pages/Send/useL1CanonicalBridge.ts:
```diff
@@ -0,0 +1,103 @@
+        return setUl1cb(true)
+      }
+
+      if (sourceTokenAmount.lte(estimatedReceived)) {
```
yeah, the logic in this hook ignores that case and only changes to the hop bridge if the price is better. otherwise we proceed with the user's choice

kangarang commented on packages/sdk/src/CanonicalBridge.ts:
```diff
@@ -106,10 +128,27 @@ class CanonicalBridge extends Base {
     let spender = this.getL1CanonicalBridgeAddress(this.tokenSymbol, chain)
-    if (chain.equals(Chain.Polygon)) {
+    if (chain.equals(Chain.Polygon) && this.tokenSymbol !== Token.MATIC) {
       spender = this.getL1PosErc20PredicateAddress(this.tokenSymbol, chain)
```
let me know if i'm mistaken, but if chain is polygon and token is not MATIC, `getL1PosErc20PredicateAddress` should return a different address than `getL1CanonicalBridgeAddress`

USDC / USDT / DAI:

```
l1CanonicalBridge: '0xA0c68C638235ee32657e8f720a23ceC1bFc77C77'
l1PosPredicate: '0x40ec5B33f54e0E8A33A975908C5BA1c14e5BbbDf'
```

ETH:

```
l1CanonicalBridge: '0xA0c68C638235ee32657e8f720a23ceC1bFc77C77'
l1PosPredicate: '0x8484Ef722627bf18ca5Ae6BcF031c23E6e922B30'
```

for MATIC we approve the canonical Plasma bridge (`0x401F6c983eA34274ec46f84D70b31C151321188b`)

shanefontaine replied:
That sounds right, but I am not 100% certain, as I'm not sure what the exact naming convention is for their bridge/token contracts. If you tested it successfully, then I believe it is good to go 👍

kangarang commented on packages/sdk/src/CanonicalBridge.ts:
```diff
@@ -139,7 +173,19 @@ class CanonicalBridge extends Base {
       )
     }
-    return l1CanonicalToken.approve(spender, amount)
+    return l1CanonicalToken.approve(spender, ethers.constants.MaxUint256)
```
normally, the app triggers a tx confirm modal which defaults to `MaxUint256`. if it's not too much work to follow the same pattern here, i'll change it back to `amount`. in my opinion, allowance txs should default to the max. metamask also allows the user to change the allowance value via `Edit Permission`.

kangarang commented on packages/sdk/src/CanonicalBridge.ts:
```diff
@@ -159,87 +217,135 @@ class CanonicalBridge extends Base {
+          amount,
+          {
+            // Gnosis requires a higher gas limit
+            gasLimit: 500e3,
```
yeahh, most of this is legacy, and i'm unsure if they are necessary anymore. if you have any insight on this, i'll defer to it!

kangarang commented on packages/sdk/src/CanonicalBridge.ts:
```diff
@@ -159,87 +217,135 @@ class CanonicalBridge extends Base {
+      }
+
+      const maxGas = 200e3
+      const gasPriceBid = '100'
```
i think these values were obtained from a tx that i populated via the arbitrum bridge. i'm unsure if they are variable. in any case, i'm going to change the values to use the ones found in [this example](https://github.com/OffchainLabs/arbitrum/blob/develop/packages/arb-ts/integration_test/arb-bridge.test.ts#L455-L462), as this is what is recommended by the arb-ts package

shanefontaine replied:
Have you tested a transaction with these values on mainnet? I know their tests have these values, but their L2 recipient or alias address might be pre-funded for these tests, which means it would work in the test but not in prod.

From a top level perspective, the user needs to pay for the bridging of the asset when using the Arbitrum bridge. This could come in the form of having funds pre-deposited on an address in L2 OR sending ETH with the L1 tx. In this case, if the `gasPriceBid` is `0`, then I think we are telling the system that we expect the funds to already exist on L2 to be paid, which I do not believe will be the case for Hop users.

It would be best to test this on the actual Arbitrum bridge itself and trace the tx and see what happens, as some of my info above might be outdated or misinformed.

Happy to call and talk through this if that is easier.

kangarang replied:
wow, it really varies. here's two successful deposits using the arbitrum bridge ui:

![Screen Shot 2022-03-07 at 9 40 25 AM](https://user-images.githubusercontent.com/18040654/157055710-743bc6bf-d3ab-4d74-b8b5-ae4ed5882f27.png)

![Screen Shot 2022-03-07 at 9 40 22 AM](https://user-images.githubusercontent.com/18040654/157055737-f410aed7-da8c-4d9f-910f-f4b6b846b834.png)

kangarang replied:
it may be best for us to integrate `arb-ts` (https://developer.offchainlabs.com/docs/frontend_integration#arbitrum--ethereum-bridge), as they suggest in their docs. it seems pretty straightforward

kangarang commented:
@shanefontaine 

> It looks like the Arbitrum bridge that the tx is sent to is not the actual address that gets approved. See attached image. It looks like the approval address for sending DAI over Arbitrum is 0xd3b5b60020504bc3489d6949d545893982ba3011, and I don't see that in the repo anywhere. I did not look at any other bridges, but it might be worth confirming that the approval addresses are accounted for if they differ from the actual bridge address that is interacted with

ahh, that makes a lot of sense! i was wondering how arbitrum handled approvals, since i had not approved the `0x72Ce9c` address. thanks for catching this.

> I thought we landed on keeping the canonical bridge logic out of the SDK and keeping the SDK pure with Hop-only logic. I see that the code is pretty closely coupled, so I'd imagine that might be difficult. Do you remember where we ended up landing on that last time we discussed this?

yeaa, i believe we all agreed that it'd be best to localize the L1 canonical bridge in the frontend. i mentioned that i wanted to get it working first, and then clean up the code after. thanks for the reminder -- i think it'll be a quick refactor

kangarang committed:
sdk/frontend: address review comments:

- use provided amount for `approve`
- use optimal _l2Gas value for optimism `depositERC20`
- use arb-ts test values for `maxGas` and `gasPriceBid`
- mv model to frontend. revert sdk changes
- rm manual gasLimit from polygon deposit txs
  * note: all other manual gas limits are necessary to trigger the tx

misc:

- add arbitrum approval spender address - l1Gateway
- capitalize dir name infoTooltip
- fix button transition and dark mode text color

shanefontaine commented on packages/sdk/src/HopBridge.ts:
```diff
@@ -428,10 +428,8 @@ class HopBridge extends Base {
-      gasLimit: 500000
-    })
+
+    return sourceChain.provider.estimateGas(populatedTx) 
```
Any insights here?

kangarang replied:
we do have a util that adds some buffer gas that we can use, but it's just a hard coded number. i feel like a percentage would indeed be more realistic and reliable.

shanefontaine commented:
@kangarang back to you. Changes look good! Left some additional comments.

It is my understanding that we will localize the addresses to the FE and effectively make no change to the `core` package before merging this PR, so I'll give one more review after that is done. Happy to talk through it if that is not what you're envisioning.

miguelmota commented on packages/frontend/src/contexts/Web3Context.tsx:
```diff
@@ -270,9 +269,13 @@ const Web3ContextProvider: FC = ({ children }) => {
   // TODO: cleanup
   const checkConnectedNetworkId = useCallback(
-    async (networkId?: number): Promise<boolean> => {
+    async (networkId?: number | string): Promise<boolean> => {
```
in what context or scenario will networkId be a `string`?

kangarang replied:
i see in many places that we convert to number just prior to calling the method. this change adds some flexibility.

![Screen Shot 2022-03-07 at 9 54 17 AM](https://user-images.githubusercontent.com/18040654/157058455-bc8fd657-989b-4a87-85a7-c38b751b5263.png)

miguelmota replied:
the `Number` wrap here is legacy code. the config `networkId` should always be a number ([type](https://github.com/hop-protocol/hop/blob/94b2570eebe7c71da7ad9d8f657d5daf0aa8dc8d/packages/core/src/networks/types.ts#L3)) and `provider.getNetwork()` will always return `number` so i don't think there's need to have string type

kangarang replied:
i see what's happening. actually, `sourceNetwork` in this particular usage is the `Network` model ([type](https://github.com/hop-protocol/hop/blob/develop/packages/frontend/src/models/Network.ts#L25)), which converts the original `networkId.toString()` ([here](https://github.com/hop-protocol/hop/blob/develop/packages/frontend/src/config/mainnet.ts#L27)). i just tested removing `.toString()` in that file, and it doesn't seem to break the app, as `Number` accepts `any`.

miguelmota replied:
nice, removing the string types is much cleaner

miguelmota commented on packages/frontend/src/pages/Send/L1CanonicalBridgeOption.tsx:
```diff
@@ -0,0 +1,60 @@
+  estimatedReceivedDisplay?: string
+  l1CanonicalBridge?: CanonicalBridge
+  sendL1CanonicalBridge?: () => void
+  setUl1cb: (u: boolean) => void
```
this callback name `setUl1cb` isn't clear what it's for, the name can be more descriptive

kangarang replied:
i'll go with `setUsingNativeBridge`

miguelmota commented on packages/core/build/addresses/mainnet.json:
```diff
@@ -34,34 +34,34 @@
     },
```
this file shouldn't be removed because [defillama](https://github.com/DefiLlama/DefiLlama-Adapters/blob/c091de607442c1430b00770c77ad2f737d6e0ed4/projects/hop/index.js#L18) relies on it (sorry it wasn't clear)

kangarang replied:
ahh, i see. are the changes here okay though?

miguelmota replied:
yes

miguelmota commented:
A few observations

expected: Select bridge with better rate after inputting amount
actual: Native bridge was not chosen when it had a better rate (maybe a debounce here is needed?)

![Peek 2022-02-16 15-52](https://user-images.githubusercontent.com/168240/154377381-c90c9e98-e615-441a-93b1-91f5b8bf7d13.gif)

expected: Disconnect wallet without error
actual: App crashed when disconnecting wallet on UI (looks like a hook somewhere needs to be updated on provider or address change) 

![Peek 2022-02-16 15-55](https://user-images.githubusercontent.com/168240/154377486-e926aaea-b524-469a-b48c-e197500484d8.gif)

miguelmota commented:
@kangarang let me know if you have any questions on the comments

kangarang committed:
fix crashed app if no signer

rename variable name

prefer cheaper bridge option.

caveat: is the user specifies a bridge, we the app will not auto-choose in the event of updated rates

use arb-ts to send L1toL2 native bridge deposits

kangarang commented:
@shanefontaine 

> Similar comment but WRT the canonical bridge addresses in core/src/addresses/

for now, let's keep the addresses in `core/src/addresses`, as the models helpers and each type relies on that single source of truth

@miguelmota 

> expected: Select bridge with better rate after inputting amount
> actual: Native bridge was not chosen when it had a better rate (maybe a debounce here is needed?)

i considered this as well, but i decided on this approach because i believe my implementation gives users more agency and less confusion.

the logic gates are: if the user has not selected either hop bridge or native bridge, the app will choose for them (whichever is cheaper). if the user explicitly chooses a particular bridge, their selection is stored in local storage and is used as the default going forward. the only other time that the app will switch bridges on behalf of the user is if the user has defaulted to native bridge, but hop's bridge is cheaper. the main benefit of this approach is that it minimizes the potential for auto-switching bridges and confusing the user. this would especially be confusing if that happened after pressing `send`.

another option is to set the bridge option at the beginning of the transaction, and prevent switching if any subsequently polled data suggest a different rate. thoughts?

miguelmota commented:
@kangarang that makes sense. the problem is that it's not clear that it'll continue using the option that was previously selected. I think there needs to be an optional setting (maybe under the cog settings) to always use the cheapest option regardless of what was previously chosen (or a way to 'reset' this state)

shanefontaine commented:
@kangarang 

> their selection is stored in local storage and is used as the default going forward

I might be misunderstanding, but I was under the impression we would always default to the cheapest option. Is that the case? _If_ the user wanted to use the more expensive option for some reason, then they'd be able to do so on a per-tx basis (not persisted in local storage).

I might be misunderstanding or not have the correct context, so let me know!

> another option is to set the bridge option at the beginning of the transaction, and prevent switching if any subsequently polled data suggest a different rate. thoughts?

I think this makes sense in general to avoid any race conditions.

kangarang commented:
@shanefontaine @miguelmota totally understand your points of views. i think a good compromise will be a one-time-only auto-bridge option set by the app (based on price) immediately after getting send data (could happen multiple times, but the app will not auto-switch beyond that point, even if a subsequent poll to get send data suggests otherwise.

kangarang commented:
working as expected:

https://user-images.githubusercontent.com/18040654/157194487-c7b9b2be-e0aa-445a-a022-7b3c0a8b75db.mp4

kangarang commented:
some notes:

bridge instantiation with `arb-ts` requires both L1 and L2 signers. due to the way our sdk configures signers/providers, injecting an Arbitrum `StaticJsonRpcProvider` prevented the bridge from performing any methods. super hard to figure out what was goin on, as the error message was the generic exception thrown by `ethers.utils.getAddress`. an actual signer was needed (i.e. connected to an address). this is resolved by instantiating an arbitrum L2 signer via `provider.getUncheckedSigner(address)`.

[this was the most complete example](https://github.com/OffchainLabs/arbitrum/blob/master/packages/arb-ts/scripts/deployStandard.ts) of `arb-ts`.

not sure if we should also be [implementing retry-ability](https://github.com/OffchainLabs/arbitrum/blob/98d33a5e92de47de97aec857c7fd92eb63db543e/packages/arb-ts/scripts/deployStandard.ts#L137)

noticed [lots of helpers](https://github.com/OffchainLabs/arbitrum/blob/master/packages/arb-ts/src/lib/bridge_helpers.ts) in `arb-ts`

miguelmota commented on packages/frontend/src/hooks/useEstimateTxCost.ts:
```diff
@@ -81,62 +94,67 @@ export function useEstimateTxCost(selectedNetwork?: Network) {
-          return gasCost
+  const estimateWrap = useCallback(async (options: EstimateTxOptions) => {
+    const { token, network } = options
+    if (!network) {
```
do we need to check that `token` exists here as well?

miguelmota commented on packages/frontend/src/hooks/useEstimateTxCost.ts:
```diff
@@ -81,62 +94,67 @@ export function useEstimateTxCost(selectedNetwork?: Network) {
+
+    try {
+      // Get estimated gas cost
+      const estimatedGasLimit = await token.wrapToken(BigNumber.from(10), true)
```
suggestion: the `true` parameter should be a variable for clarity: `const estimateOnly = true`

kangarang replied:
i did not realize we had a method `getWrapTokenEstimatedGas` for this already

miguelmota commented on packages/frontend/src/models/CanonicalBridge.ts:
```diff
@@ -0,0 +1,341 @@
+
+  constructor(network: string, signer: Signer, token: TToken, chain: TChain) {
+    super(network, signer)
+    if (!token) {
```
should all the parameters be checked here that they exists?

miguelmota commented on packages/frontend/src/models/CanonicalBridge.ts:
```diff
@@ -0,0 +1,341 @@
+    return this.signer.sendTransaction(populatedTx)
+  }
+
+  public async populateDepositTx(amount: BigNumberish, customRecipient?: string): Promise<any> {
```
suggestion: instead of multiple if statements checking chain and token, using polymorphism would be cleaner ([example](https://github.com/hop-protocol/hop/blob/develop/packages/sdk/src/watchers/Watcher.ts))

miguelmota commented on packages/frontend/src/pages/Convert/ConvertOption/HopConvertOption.tsx:
```diff
@@ -68,11 +67,12 @@ class HopConvertOption extends ConvertOption {
+      amountIn,
+      sourceNetwork.slug,
+      destNetwork.slug,
+      true
```
suggestion: move boolean parameters to variables for clarity

miguelmota commented on packages/sdk/src/Base.ts:
```diff
@@ -373,36 +388,27 @@ class Base {
         if (connectedChainId !== chain.chainId) {
-          if (!signer.provider) {
-            // console.log('connect provider')
-            return (signer as Signer).connect(chain.provider)
```
curious, why was this line removed?

kangarang replied:
this block of code shouldn't ever run because it is within the condition set by line 382

miguelmota replied:
good catch
i think we should still connect `chain.provider` if `signer.provider` doesn't exist since it's possible to use in a signer without a provider and let the sdk handle connecting provider

kangarang replied:
yes, thank you, very thoughtful. on line [409](https://github.com/hop-protocol/hop/blob/796bdf3d106b521579f998a86a7fcd718bdc426c/packages/sdk/src/Base.ts#L409), we connect `chain.provider` as a fallback

miguelmota replied:
This should return the signer with connected provider instead of just provider, otherwise transaction won't be able to be made with just the provider

```typescript
return signer.connect(chain.provider)
```

kangarang replied:
ahhhh, i see. thanks for clarifying

miguelmota commented on packages/frontend/src/pages/Send/useL1CanonicalBridge.ts:
```diff
@@ -0,0 +1,136 @@
+import { handleTransaction } from './useSendTransaction'
+import { useTransactionReplacement } from 'src/hooks'
+
+async function needsApproval(l1CanonicalBridge: CanonicalBridge, sourceToken, sourceTokenAmount) {
```
the first parameter has a type but the other two do not. I think they should all have types

miguelmota commented on packages/frontend/src/pages/Send/useL1CanonicalBridge.ts:
```diff
@@ -0,0 +1,136 @@
+
+    const sourceNetwork = findNetworkById(sourceToken?.chain.chainId.toString()!)
+
+    if (await needsApproval(l1CanonicalBridge, sourceToken, sourceTokenAmount)) {
```
for best practice, the `await` should be moved out of the condition into a var (ie `shouldApprove`)

miguelmota commented on packages/frontend/src/pages/Send/useL1CanonicalBridge.ts:
```diff
@@ -0,0 +1,136 @@
+      },
+      onConfirm: async () => {
+        try {
+          const isNetworkConnected = await checkConnectedNetworkId(1)
```
we should avoid using hard coded chain ids since it will make it harder to make it work with a testnet. suggestion: use the util to get L1 network id since that reads the network ids from config

miguelmota commented on packages/frontend/src/pages/Send/useSendData.tsx:
```diff
@@ -31,7 +31,7 @@ const useSendData = (
       enabled:
         !!token?.address && !!fromNetwork?.slug && !!toNetwork?.slug && !!fromAmount?.toString(),
-      refetchInterval: 5e3,
+      refetchInterval: 10e3,
```
it looks like the intervals vary in components; ie `5`, `7`, or `10` seconds. maybe this can be more consistent throughout

miguelmota commented on packages/frontend/src/models/CanonicalBridge.ts:
```diff
@@ -0,0 +1,341 @@
+} from '@hop-protocol/core/contracts'
+import { ChainSlug, CanonicalToken } from '@hop-protocol/sdk'
+import Base, { L1Factory } from '@hop-protocol/sdk/dist/src/Base'
+import { metadata } from '@hop-protocol/sdk/dist/src/config'
```
I don't think it should be importing from `dist` like this `@hop-protocol/sdk/dist/src/config` because that makes it fragile. I suggest importing from `@hop-protocol/sdk` or structuring the typescript build so the output dirs are more structured like `@hop-protocol/sdk/config`

miguelmota commented on packages/frontend/src/models/CanonicalBridge.ts:
```diff
@@ -0,0 +1,341 @@
+  }
+
+  public getL1CanonicalAllowance(): Promise<BigNumber> {
+    const l1CanonicalToken = this.getL1Token().connect(this.signer)
```
suggestion: move `.connect(this.signer)` be to `getL1Token` method to avoid having to manually connect every time `getL1Token` is called

miguelmota commented on packages/frontend/src/models/CanonicalBridge.ts:
```diff
@@ -0,0 +1,341 @@
+            l1CanonicalToken.address,
+            l2TokenAddress,
+            amount,
+            BigNumber.from('1920000'),
```
this should be moved to a variable to know what it means

miguelmota commented on packages/frontend/src/models/CanonicalBridge.ts:
```diff
@@ -0,0 +1,341 @@
+            l2TokenAddress,
+            amount,
+            BigNumber.from('1920000'),
+            '0x',
```
this should be moved to a variable to know what it means

miguelmota commented on packages/frontend/src/models/CanonicalBridge.ts:
```diff
@@ -0,0 +1,341 @@
+        } else {
+          const fee = price.mul(gasNeeded)
+          if (fee.gt(walletBal)) {
+            console.log(
```
does metamask throw error in this case?

kangarang replied:
that is a good question. i haven't run into this issue yet, and this is copy/pasta from an [arb-ts example](https://github.com/OffchainLabs/arbitrum/blob/338fa2da63c290faeaa27b848f1d93d60038aa09/packages/arb-ts/scripts/deployStandard.ts#L121)

kangarang committed:
address review comments

set networkId type: number

kangarang reviewed (commented):
just a couple more edits and i think we should be good to merge. thanks for the awesome code reviews

kangarang committed:
update txConfirm args

kangarang committed:
add contract interfaces

miguelmota commented:
Code updates look good 👍 

can we deploy to a staging site (mainnet2) for live testing?

kangarang committed:
add L1_OptimismGateway abi interface

kangarang committed:
cleanup. update lockfile

kangarang committed:
improve allowance hook

kangarang committed:
improve native bridge deposits

kangarang commented on packages/frontend/src/config/nativeTokenBridges.ts:
```diff
@@ -0,0 +1,36 @@
+  L1XDaiWETHOmnibridgeRouter__factory,
+} from '@hop-protocol/core/contracts'
+
+export const nativeTokenBridges = {
```
this isn't being used atm, but will be useful in a subsequent refactor

kangarang committed:
address review comment

kangarang commented:
i've tested each native bridge 3 times: (1) ETH, (2) native asset, and (3) DAI. where 2 of the 3 are the same, I sent USDC (i.e. not polygon).

the final few commits includes fixes where [such transfers failed](https://etherscan.io/address/0x9997da3de3ec197c853bcc96caecf08a81de9d69)

miguelmota commented:
minor bug 

expected: when wallet is not connected, send button should prompt to connect wallet or be disabled with message to connect wallet
actual: send button is active but nothing happens

![DeepinScreenshot_select-area_20220325093046](https://user-images.githubusercontent.com/168240/160162102-805b7a21-f908-47a6-82f2-a51d0f020c1b.png)

pools page send button for reference

![DeepinScreenshot_select-area_20220325093230](https://user-images.githubusercontent.com/168240/160162389-66a597e7-172d-4846-98ad-e928538b91a3.png)

miguelmota commented:
bug

expected: ability to approve or send tokens, or show message why buttons are disabled
actual: all action buttons are disabled

approve and send buttons also appear to be different sizes

![DeepinScreenshot_select-area_20220325093455](https://user-images.githubusercontent.com/168240/160162744-5fd691c0-227f-415d-a7a4-f7bf2e4cefdc.png)

recording

![Peek 2022-03-25 09-37](https://user-images.githubusercontent.com/168240/160163139-5909ba52-0845-4d46-ab8b-d6e44ae9ed19.gif)

miguelmota commented:
bug

expected: show warning and disable button if account doesn't have enough balance to send
actual: no warning is shown and send button is enabled but transaction fails leaving ui in a stuck state

![DeepinScreenshot_select-area_20220325093912](https://user-images.githubusercontent.com/168240/160163426-52e85727-f72d-4445-bb77-76b338ddc5bb.png)

recording

![Peek 2022-03-25 09-39](https://user-images.githubusercontent.com/168240/160163571-ceb2e99f-a79e-45c4-969f-42ff67ecf2fb.gif)

kangarang committed:
add cors headers

kangarang committed:
reduce wait time for approval txs

kangarang committed:
add disable check for unconnected wallet

kangarang committed:
cleanup:

- add else if bnc wallet includes `connect()`
- cleanup web3context
- fail gracefully when MM throws
- mv needsNativeBridgeApproval -> use approval hook
- manually set gas limits for polygon txs
- re-org  send page so that approval can happen
- add a tx modal to approval txs
- wrap getCode in try-catch

kangarang committed:
use same width for approval and send buttons

kangarang committed:
rm loading from approval button

kangarang committed:
fix stats page data population

kangarang committed:
improve native bridge checks:

note: enabled-button bug is still present when user has not approved bridge

kangarang committed:
stash: manually set arb args

miguelmota commented:
bug

expected: ability to approve or send tokens, or show message why buttons are disabled
actual: all action buttons are disabled without a reason message

![Peek 2022-04-11 11-43](https://user-images.githubusercontent.com/168240/162807950-798092ef-7945-43f8-863d-9e0e31b97085.gif)

kangarang commented:
@miguelmota, thank you for your thorough bug reports. i have resolved these issues in my latest changes locally, but i need to do some further testing.

<img width="961" alt="Screen Shot 2022-04-12 at 12 43 36 PM" src="https://user-images.githubusercontent.com/18040654/163012683-fbf7278b-5dae-494e-9bb1-56cbe3886e62.png">

will push fixes shortly this afternoon

kangarang committed:
add addresses to be filtered to native bridges

kangarang committed:
fix txConfirm modal. simplify approval hook

kangarang committed:
add dep to chk if ntv bridge approval is necessary

kangarang committed:
cleanup

kangarang committed:
simplify spenders

kangarang committed:
update lockfile

kangarang committed:
progress...

kangarang committed:
rename network -> chain

kangarang committed:
refactor Send component:

previously, the Send component was over 750 lines
and very difficult to reason about.

this commit shortens the Send component to < 600 lines,
reducing some of the mental overhead by modularizing the following:
- chain selection
- converting values displayed
- warning, error, and info messaging
- moving tx estimations into appropriate hooks and converting them to
  callback hooks with deps
- continue renaming network -> chain

kangarang committed:
replace old Send component

kangarang committed:
fix addresses, simplify canonical bridges library

kangarang committed:
separate setting source and destination providers

kangarang committed:
rm extraneous code

kangarang committed:
improve flow of approving/sending native deposit

- cleanup

kangarang committed:
add fetch headers

kangarang committed:
improve approval txs

kangarang commented:
Known bug: other routes than `/send` are no longer working as expected. seems to have something to do with the `txConfirm` changes

shanefontaine commented:
Replaced with #435
